Add `--preserve-filetime` argument to support persisting file modification time

desrosj opened the pull request:
As detailed in #41, preserving the file time when importing local files could be very useful.

This PR introduces the `--preserve-filetime` option to use the file modified time instead of the default current time.

This is missing tests as I have not written Behat tests before. I will add those shortly as soon as I can read up a bit and get them running locally.

Fixes #41

schlessera commented:
As discussed in Slack, please use both creation and modification dates, and make sure to calculate GMT difference correctly.

danielbachhuber commented on src/Media_Command.php:
```diff
@@ -183,6 +183,9 @@ function regenerate( $args, $assoc_args = array() ) {
 	 * : Output just the new attachment ID.
 	 *
+	 * [--filetime]
+	 * : Use the file modified time as the post date. Only works if the file is local.
```
Is there any reason not to make this the default behavior?

schlessera replied:
It would technically be a breaking change, but it would certainly make sense. Not sure anyone would rely on the old behaviour.

desrosj replied:
The only argument I can think of against making this the default behavior is that it will only work when the files are local.

schlessera commented on src/Media_Command.php:
```diff
@@ -269,11 +281,19 @@ function import( $args, $assoc_args = array() ) {
 			);
+
+			if ( ! empty( $file_time ) ) {
+				$post_array['post_date'] = gmdate( 'Y-m-d H:i:s', $file_time + ( $gmt_offset * 60 * 60 ) );
```
Use the constant `HOUR_IN_SECONDS` here instead of a magic number.

schlessera commented on src/Media_Command.php:
```diff
@@ -269,11 +281,19 @@ function import( $args, $assoc_args = array() ) {
+			if ( ! empty( $file_time ) ) {
+				$post_array['post_date'] = gmdate( 'Y-m-d H:i:s', $file_time + ( $gmt_offset * 60 * 60 ) );
+				$post_array['post_date_gmt'] = gmdate( 'Y-m-d H:i:s', $file_time );
+				$post_array['post_modified'] = gmdate( 'Y-m-d H:i:s', $file_time + ( $gmt_offset * 60 * 60 ) );
```
Use the constant `HOUR_IN_SECONDS` here instead of a magic number.

schlessera commented:
Small nitpick regarding constants vs magic numbers.

Are you up to writing tests for this as well?

desrosj commented:
Updated to use `HOUR_IN_SECONDS` instead.

I added what I have so far for tests. I am missing something in the `touch` step causing the file modified time to not be correct when tested. Also, I'm not sure if I have included more `And` steps then accepted in my test.

I looked into the file created and modified times. The times should all be returned as Unix timestamps that can be converted with `data()`. Also, there is no created time for files in most Unix filesystems.

`filemtime()` will return the `the time when the data blocks of a file were being written to, that is, the time when the content of the file was changed` while `filectime()` returns when `the inode data is changed`. `filectime()` will most likely return a newer date than `filemtime()` in most cases because of that difference. My patch is using the `filemtime()` for published and modified dates. If there is a better method, I can update accordingly.

danielbachhuber commented on features/media-import.feature:
```diff
@@ -60,6 +60,27 @@ Feature: Manage WordPress attachments
+      | {CACHE_DIR}/large-image.jpg | http://wp-cli.org/behat-data/large-image.jpg     |
+    And I run `touch -m 1505787257 {CACHE_DIR}/large-image.jpg`
+    And I run `wp option update gmt_offset -5`
+    And I run `wp media import {CACHE_DIR}/large-image.jpg --post_id=1 --filetime`
```
This should be `When I run`, and include a `Then STDOUT should contain` statement

danielbachhuber commented on features/media-import.feature:
```diff
@@ -60,6 +60,27 @@ Feature: Manage WordPress attachments
+    And I run `touch -m 1505787257 {CACHE_DIR}/large-image.jpg`
+    And I run `wp option update gmt_offset -5`
+    And I run `wp media import {CACHE_DIR}/large-image.jpg --post_id=1 --filetime`
+    And I run `wp post get 1 --field=post_date`
```
This should be `When I run`

danielbachhuber commented on features/media-import.feature:
```diff
@@ -60,6 +60,27 @@ Feature: Manage WordPress attachments
+      2000-01-01 00:00:00
+      """
+
+    And the return code should be 0
```
This `the return code...` statement is superfluous

danielbachhuber commented on src/Media_Command.php:
```diff
@@ -183,6 +183,10 @@ function regenerate( $args, $assoc_args = array() ) {
 	 * [--porcelain]
 	 * : Output just the new attachment ID.
 	 *
+	 * [--filetime]
```
Can we make this flag more specific? Possibly `--use-filetime-for-date`?

See https://make.wordpress.org/cli/handbook/philosophy/#readability-trumps-number-of-keystrokes

danielbachhuber commented:
@desrosj Do you need assistance finishing these changes up?

danielbachhuber commented:
@desrosj ^ worth taking a look through the cleanup I did when you have a moment.

desrosj commented:
@danielbachhuber Thanks! Checked out your updates. Just didn't have the time the last week to follow up with this.
